package/upgrade/lib.sh

```diff
@@ -317,3 +317,144 @@ shutdown_all_vms()
+  helm history $release_name -n $namespace || true
+
+  local cnt=$(yq e '. | length' $tmp_file)
+  if [ "$cnt" -gt 0 ]; then
+    local current_status=$(yq e '.[-1].status' $tmp_file)
+    if [ "$current_status" == "pending-upgrade" ]; then
+
+      # in this state at least 180 seconds, to avoid clash with the normal pending upgrade
+      local dt1=$(yq e '.[-1].updated' $tmp_file)
+      dt1=$(date -d "$dt1" +%s)
+      local dt2=$(date +%s)
+      local delta=$(( dt2 - dt1 ))
+      if [ ! $delta -gt 180 ]; then
+        echo "the release is in pending-upgrade state for $delta seconds, do not rollback"
+        rm -f $tmp_file
+        return
+      fi
+
+      if [ "$cnt" -gt 1 ]; then
+        local previous_status=$(yq e '.[-2].status' $tmp_file)
+        if [ "$previous_status" == "deployed" ]; then
+          local previous_revision=$(yq e '.[-2].revision' $tmp_file)
+          echo "rollback pending release from"
+          yq e '.[-1]' $tmp_file
+          echo "to"
+          yq e '.[-2]' $tmp_file
+
+          helm rollback -n $namespace $release_name $previous_revison || EXIT_CODE=$?
+          if [ $EXIT_CODE != 0 ]; then
+            echo "helm rollback $release_name -n $namespace failed $EXIT_CODE, try later"
+          fi
+
+          echo "check release history after rollback"
+          helm history $release_name -n $namespace || true
+        fi
+      else
+        echo "the $release_name is pending-upgrade, but there is no previous revision to rollback"
+      fi
+    else
+      echo "the $release_name is $current_status"
+      echo "when the helm release has been rolled back, the fleet bundle takes time to be ready, wait"
+    fi
+  else
+    echo "there is no helm history of $release_name in $namespace"
+  fi
+  rm -f $tmp_file
+}
+
+check_fleet_bundle_helm_releases()
+{
+  local tmp_file=/tmp/bundle_status.tmp.txt
+  local EXIT_CODE=0
+  # the bundle object is too big, fetch the wanted information only
+  kubectl get bundle -A -o=custom-columns='NAME:.metadata.name,HELM_NAME:.spec.helm.releaseName,NAMESPACE:.spec.defaultNamespace,MESSAGE:.status.conditions[?(@.type=="Ready")].message' --no-headers > $tmp_file || EXIT_CODE=$?
+
+  if [ $EXIT_CODE != 0 ]; then
+    echo "kubectl get bundle failed $EXIT_CODE, try later"
+    return
+  fi
+
+  while read -r name helm_name ns message; do
+    if [[ $message == *"another operation"* ]]; then
+      echo "bundle $name is pending_upgrade: $message"
+      if [[ $helm_name == "<none>" ]]; then
+        rname=$name
+      else
+        rname=$helm_name
+      fi
+      echo "prepare to rollback $rname in $ns"
+      try_rollback_helm_release $ns $rname
+    fi
+  done < $tmp_file
+  rm -f $tmp_file
+}
+
+wait_for_fleet_bundles()
+{
+  # wait for the changes to be applied to bundle object first
+  # fleet-agent takes time to further apply changes to downstream deployments
+  sleep 10
+  if [ -z "${HARVESTER_UPGRADE_HELM_AUTOMATIC_ROLLBACK}" ]; then
```

Comment:
I feel it might be redundant to manage the rollback via the upgrade script, as it will be better for users to manually check the helm status and roll it back one by one where applicable.

Reply:
the `helm rollback` is an advanced operation, user may not be confident/careful to do all the related cheks and operations, the code can do the rollback when some specific conditons are met

the code is left for future extension, it depends on the feedback from users, when they are requesting automaic rollback, then we can start from here

Reply:
just a preference that less code is easier to understand and maintain over time, we can keep it in our own branch and bring it back when necessary, thanks.

Reply:
removed code, thanks.